File: src/main/java/ru/vood/freemarker/ext/sql/FtlDefaultObjectWrapper.java
package ru.vood.freemarker.ext.sql;

import freemarker.template.TemplateModel;
import freemarker.template.TemplateModelException;
import freemarker.template.Version;
import ru.vood.freemarker.ext.sql.model.*;

import java.sql.Array;
import java.sql.Clob;
import java.sql.Struct;

public class FtlDefaultObjectWrapper extends freemarker.template.DefaultObjectWrapper {


    public FtlDefaultObjectWrapper(Version incompatibleImprovements) {
        super(incompatibleImprovements);
    }


    protected TemplateModel handleUnknownType(Object obj) throws TemplateModelException {
        if (obj instanceof Array) {
            return new ArrayModel((Array) obj, this);
        }
        if (obj instanceof Clob) {
            return new ClobModel((Clob) obj, this);
        }
        if (obj instanceof Struct) {
            return new StructModel((Struct) obj, this);
        }
        if (obj instanceof FetchedResultSet) {
            return new FetchedResultSetModel((FetchedResultSet) obj, this);
        }
        if (obj instanceof FetchedResultSetTransposed) {
            return new FetchedResultSetTransposedModel((FetchedResultSetTransposed) obj, this);
        }
        return super.handleUnknownType(obj);
    }


}

File: src/main/java/ru/vood/freemarker/ext/sql/FtlDefaultObjectWrapper.kt
package ru.vood.freemarker.ext.sql

import freemarker.template.DefaultObjectWrapper
import freemarker.template.TemplateModel
import freemarker.template.Version
import ru.vood.freemarker.ext.sql.model.*
import java.sql.Array
import java.sql.Clob
import java.sql.Struct

/**
 * The FTLDB's default object wrapper. Extends FreeMarker's wrapper with SQL type wrapping.
 *
 *
 * Registered wrappers are:
 *
 *  * [ArrayModel] - treats SQL collections ([Array]) as sequences
 *  * [ClobModel] - treats CLOBs ([Clob]) as strings
 *  * [StructModel] - treats UDTs ([Struct]) as sequences of elements
 *  * [FetchedResultSetModel] - treats fetched result sets ([FetchedResultSet]) as 2-layer objects
 *  * [FetchedResultSetTransposedModel] - wraps transposed result sets
 *
 */
class FtlDefaultObjectWrapper(incompatibleImprovements: Version?) : DefaultObjectWrapper(incompatibleImprovements) {

    override fun handleUnknownType(obj: Any): TemplateModel {
        return when (obj) {
            is Array -> ArrayModel(obj, this)
            is Clob -> ClobModel(obj, this)
            is Struct -> StructModel(obj, this)
            is FetchedResultSet -> FetchedResultSetModel(obj, this)
            is FetchedResultSetTransposed -> FetchedResultSetTransposedModel(obj, this)
            else -> super.handleUnknownType(obj)
        }

    }
}